src/sentry/services/hybrid_cloud/audit/__init__.py

```diff
@@ -0,0 +1,78 @@
+from __future__ import annotations
+
+import abc
+from dataclasses import dataclass
+from typing import Any, Mapping
+
+from sentry.services.hybrid_cloud import InterfaceWithLifecycle, silo_mode_delegation, stubbed
+from sentry.services.hybrid_cloud.auth import ApiAuthIdentity, ApiAuthProvider
+from sentry.services.hybrid_cloud.organization import ApiOrganization, ApiOrganizationMember
+from sentry.services.hybrid_cloud.user import APIUser
+from sentry.silo import SiloMode
+
+
+@dataclass
+class AuditLogMetadata:
+    organization: ApiOrganization
+    event: int
+    actor_label: str | None = None
+    actor: APIUser | None = None
+    ip_address: str | None = None
+    target_object: int | None = None
+    target_user: APIUser | None = None
+
+
+class AuditLogService(InterfaceWithLifecycle):
+    @abc.abstractmethod
+    def write_audit_log(
+        self,
+        *,
+        metadata: AuditLogMetadata,
+        data: Mapping[str, Any],
+    ) -> None:
+        raise NotImplementedError
+
+    @abc.abstractmethod
+    def log_organization_membership(
+        self,
+        *,
+        metadata: AuditLogMetadata,
+        organization_member: ApiOrganizationMember,
+    ) -> None:
+        raise NotImplementedError
```

Comment:
Are you thinking that there will be a method for each audit log that is recorded? We have quite a few audit log event types which will make this interface wide.

We should be able to extend the existing AuditLogEvent plumbing code to be aware of required data attributes which would let us have an interface like

```python
audit_log_service.log(AuditLogMetaData( 
    organization=self.organization,
    actor=self.user,
    ip_address=self.request.META["REMOTE_ADDR"],
    target_object=auth_identity.id,
    event=audit_log.get_event_id("SSO_IDENTITY_LINK"),
    auth_identity=auth_identity
)
```
With required parameters embedded in to the `event` definition `AuditLogMetaData` could ensure all the required attributes are present. This approach would have a narrow interface, but wouldn't be as typesound as what you have presently.